Source/JobGiver_ZombieResponse.cs

```diff
@@ -13,87 +13,61 @@ public class JobGiver_ZombieResponse : ThinkNode_JobGiver
         // Token: 0x060003E6 RID: 998 RVA: 0x0002923C File Offset: 0x0002763C
         protected override Job TryGiveJob(Pawn pawn)
         {
-            return this.TryGetAttackNearbyEnemyJob(pawn);
+            Pawn_Zombiefied predator = pawn as Pawn_Zombiefied;
+            Pawn prey = BestPawnToHuntForPredator(predator);
+            if (prey == null) return null;
+            return new Job(ZombiefiedMod.zombieHunt, prey);
         }
 
-        // Token: 0x060003E7 RID: 999 RVA: 0x000292AC File Offset: 0x000276AC
-        private Job TryGetAttackNearbyEnemyJob(Pawn pawn)
-        {
-            Pawn thing = BestPawnToHuntForPredator(pawn);
-            if (thing != null)
-            {
-                return new Job(ZombiefiedMod.zombieHunt, thing)
-                {
-                    killIncappedTarget = true,
-                    expiryInterval = (int)(Rand.RangeSeeded(1f, 2f, Find.TickManager.TicksAbs) * 700),
-                    attackDoorIfTargetLost = true
-                };
-            }
-            return null;
-        }
 
-        public Pawn BestPawnToHuntForPredator(Pawn predator, float range = 7f)
+        public Pawn BestPawnToHuntForPredator(Pawn_Zombiefied predator, float range = 7f)
         {
-            //List<Pawn> allPawnsSpawned = predator.Map.mapPawns.AllPawnsSpawned;
-            List<Thing> allThingsRegion = predator.GetRegion().ListerThings.AllThings;
+            List<Thing> allThingsRegion = predator.Map.listerThings.AllThings;
 
             Pawn pawnToReturn = null;
-            float num = 0f;
+            float closest = 0f;
 
-            //for (int i = 0; i < allPawnsSpawned.Count; i++)
-            for (int i = 0; i < allThingsRegion.Count; i++)
-                {
-                Pawn pawn2 = allThingsRegion[i] as Pawn;
-                if (pawn2 != null && predator != pawn2)
+            foreach (Thing item in allThingsRegion)
+            {
+                Pawn prey = item as Pawn;
+
+                // if prey is a pawn or the predator, skip.
+                if (prey == null || predator == prey) continue;
+                // if prey isn't acceptable, skip.
+                if (!IsAcceptablePreyFor(predator, prey)) continue;
+                // if prey is too far, skip.
+                float distance = GetDistance(predator, prey);
+                if (distance > range) continue;
+                // if predator can't reach the prey, skip.
+                if (!predator.CanReach(prey, ZombiefiedMod.allowBreaching)) continue;
+                // if predator can reach prey, but it has to go through walls, add to distance, to ensure un-blocked pawns are considered closer, and thus easier to get to
+                if (ZombiefiedMod.allowBreaching && !predator.CanReach(prey, false)) distance *= 5;
```

Comment:
You dont want to call CanReach twice if avoidable because pathing is expensive

Reply:
Hmm. Got an idea for how to do it? In this case we only do it twice if breaching is allowed, we have to do it twice to figure out if it can reach it without going through walls, and with going through walls.

Reply:
Probably just check for distance and if for some reason it cant reach the prey eventho it can destroy anything in its path just let the job fail

Reply:
That should be doable. I guess this check is only to consider pawns behind walls harder to get to, to get the zombie to get something without trying to breach all the time.

What do you think about having this check behind a configuration just in case? Default it to off, people can turn it on if they want to, warn em about a performance increase because if it or something

Reply:
I dont mind but then noone is ever gonna use it. People dont ususally check options. Also theyre gonna enable it thinking they wouldnt care about the performance hit and then theyre still gonna hate the performance hit

Reply:
But I also think this can work out well enough to be included by default just needs some work: Performance needs to be good, there needs to be a reason for them to attack through walls (maybe embrasures or sounds from inside the walls).
Right now they just have wallhack as you called it. that is kinda weird but has potential to be neat

Reply:
> That should be doable. I guess this check is only to consider pawns behind walls harder to get to, to get the zombie to get something without trying to breach all the time.

```
if(!predator.CanReach(prey, false))
{
   if(ZombiefiedMod.allowBreaching)
   {
      distance *=5;
   } else {
      continue;
   }
}
```

Reply:
>there needs to be a reason for them to attack through walls (maybe embrasures or sounds from inside the walls)

Totally want to work out some reasoning for it yeah. I want to look into the fog of war mod and figure out what they did to figure out which buildings are able to be seen through. Embrasures are, for example, and I imagine mods that add glass probably are too. I figure zombies wanting to path through those types of things makes sense.

Right now though, it's terrifying, and certainly an increase in difficulty.

Reply:
Might also just check if embrasures exist in the game (so an embrasures mod is active) and enable allowBreaching if that is the case

Reply:
Also might allowBreaching only when moving towards a noisy location, not when hunting.

Just ideas btw

Reply:
Or maybe allowBreaching only when moving towards a noisy location and then they also find a prey behind walls

Reply:
>``` if(!predator.CanReach(prey, false))
> {
>    if(ZombiefiedMod.allowBreaching)
>    {
>       distance *=5;
>    } else {
>       continue;
>    }
> }```

That'd consider any pawn they can reach without going through a wall *=5 though.

> Or maybe allowBreaching only when moving towards a noisy location and then they also find a prey behind walls

Oooh. Hmm. well, they already try to get to noisy locations, so that part is handled, and I've noticed they tend to gather at walls. Maybe we can just have breaching only happen when they're nearby already? 🤔Sounds super doable, would just require putting a distance check on the blocking buildings toil

I think if it's on when doing noisy locations though, I would think that'd mean they'd more-regularly bust down walls if you have a wall around your houses. At least with the hunting thing they've gotta be in your general vicinity.